pytorch_lightning/callbacks/base.py

```diff
@@ -282,24 +282,52 @@ def on_keyboard_interrupt(self, trainer: "pl.Trainer", pl_module: "pl.LightningM
     def on_exception(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", exception: BaseException) -> None:
         """Called when any trainer execution is interrupted by an exception."""
 
+    def state_dict(self) -> Dict[str, Any]:
+        """Called when saving a checkpoint, implement to generate callback's ``state_dict``.
+
+        Returns:
+            A dictionary containing callback state.
+        """
+        return {}
+
+    def load_state_dict(self, state_dict: Dict[str, Any]) -> None:
+        """Called when loading a checkpoint, implement to reload callback state given callback's ``state_dict``.
+
+        Args:
+            state_dict: the callback state returned by ``state_dict``.
+        """
+        pass
+
     def on_save_checkpoint(
         self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", checkpoint: Dict[str, Any]
-    ) -> dict:
-        """Called when saving a model checkpoint, use to persist state.
+    ) -> Optional[dict]:
+        r"""
+        .. deprecated:: v1.6
+            Returning state in this callback hook was deprecated in v1.6 and will be removed in v1.8.
+            Use ``state_dict`` instead to return state. In v1.8 ``on_save_checkpoint`` can only return None.
+
+        Called when saving a checkpoint to give you a chance to store anything else you might want to save.
 
         Args:
             trainer: the current :class:`~pytorch_lightning.trainer.Trainer` instance.
             pl_module: the current :class:`~pytorch_lightning.core.lightning.LightningModule` instance.
             checkpoint: the checkpoint dictionary that will be saved.
 
         Returns:
-            The callback state.
+            None or the callback state. Support for returning callback state will be removed in v1.8.
         """
 
     def on_load_checkpoint(
         self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", callback_state: Dict[str, Any]
     ) -> None:
-        """Called when loading a model checkpoint, use to reload state.
+        r"""
+        .. deprecated:: v1.6
+            This callback hook was deprecated in v1.6 and will be removed in v1.8. Use
+            ``load_state_dict`` instead. In v1.8 ``on_load_checkpoint(checkpoint)`` will replace
+            ``on_load_checkpoint(callback_state)``, i.e. input argument will be entire loaded
```

Comment:
I'm confused by this message.
We won't remove the hook. What's deprecated is the signature only. If I understand correctly from the description, you will do the signature inspection + deprecation warning in a follow up PR. 

Notice: The deprecation (including all multi-part PRs) has to be completed before 1.6 or has to be dropped from 1.6 and added in 1.7.

Reply:
> We won't remove the hook. What's deprecated is the signature only. If I understand correctly from the description, you will do the signature inspection + deprecation warning in a follow up PR.

correct. do we want the deprecation annotations here? the different signatures will become BC so I added the annotation to raise more visibility

Reply:
Thanks for the confirmation. So I'm requesting to change:

> This callback hook was deprecated in v1.6 and will be removed in v1.8.

This cannot say that the hook gets removed. The hook will stay (the name is the same) but its signature and behavior will. 
IMO, a message like this would be more appropriate:

> This callback hook will change its signature and behavior in v1.8. 
> If you wish to load the state of the callback, use ``load_state_dict`` directly.
> In v1.8, ``on_load_checkpoint(checkpoint)`` will receive the entire loaded ....

Reply:
ahh got it, agree this message is better! updated

Reply:
per below, leaving the `on_save/load_checkpoint` changes for a separate PR